### ISSUE TITLE
Sub-Instances of Hono don't trigger onError

### ISSUE THREAD

**Issue by nanderLP** [honojs/hono#851]
Hi!
I'm currently restructuring my project's routes to use multiple sub-instances for their respective category. Hono supports grouping multiple instances with [`hono.route()`](https://honojs.dev/docs/api/routing/#grouping), though it seems like it's not possible to define a custom error handler for a sub-instance.  
  
Here is my example code:
```ts
import { Hono } from "hono";
import { serve } from "https://deno.land/std/http/server.ts";

const hono = new Hono();

const route = new Hono();

route.onError((err, ctx) => {
  return ctx.text("Caught error!", 500);
});

route.get("/", (ctx) => {
  throw new Error("Unhandled Error.");
});

hono.route("/", route);

serve(hono.fetch, { port: 3000 });
```

The endpoint `/`, defined by `route`, will throw an Error. The instance `route` *should* catch and handle errors, because of `route.onError(...)`.  
Though when hitting the endpoint `/`, the thrown error gets printed to the console and the API returns `500 - Internal Server Error`.
```
Trace: Error: Unhandled Error.
    at file:///workspace/src/server.ts:13:9
    at Hono.dispatch (https://deno.land/x/hono@v2.7.5/hono.ts:216:15)
    at Server.fetch (https://deno.land/x/hono@v2.7.5/hono.ts:263:17)
    at Server.#respond (https://deno.land/std@0.174.0/http/server.ts:299:37)
    at Server.#serveHttp (https://deno.land/std@0.174.0/http/server.ts:346:20)
    at Hono.errorHandler (https://deno.land/x/hono@v2.7.5/hono.ts:111:13)
    at Hono.handleError (https://deno.land/x/hono@v2.7.5/hono.ts:191:19)
    at Hono.dispatch (https://deno.land/x/hono@v2.7.5/hono.ts:219:21)
    at Server.fetch (https://deno.land/x/hono@v2.7.5/hono.ts:263:17)
    at Server.#respond (https://deno.land/std@0.174.0/http/server.ts:299:37)
    at Server.#serveHttp (https://deno.land/std@0.174.0/http/server.ts:346:20)
```  
This looks like a bug to me. I appreciate any help!

**Comment by nanderLP** [comment]
Seems like the problem is caused by the implementation of `hono.route()`.
https://github.com/honojs/hono/blob/a9746b2296f38f740a64c32eeacfdaa9c568245e/src/hono.ts#L117-L126
`hono.route()` only imports the routes and leaves the rest untouched. Is this intended behavior? I feel like it's bad DX to support a custom error handler, when it won't be used in this case.

**Comment by yusukebe** [comment]
Hi @nanderLP !

This is not a bug. This is because `hono.route('/', route)` is designed to pass the routings of `route` to `hono`. The `onError` handler is not included in the routing. So, `onError` in `hono` will be used for any route.

> I feel like it's bad DX to support a custom error handler, when it won't be used in this case.

I think you are right. It is very difficult to make this possible, but worth a try. Please keep this Issue open for a while. Thanks.

**Comment by yusukebe** [comment]
By the way, similarly, `route.noFound()` is not used, so that is also the same matter.

**Comment by yusukebe** [comment]
Hi @usualoma 

Do you have any good ideas?

One of mine is making a route that includes the error handler and the not found handler, not only the handler.  The definitions will be like the following:

```ts
type HandlerContainer = {
  handler: Handler
  errorHandler: ErrorHandler | undefined
  notFoundHandler: NotFoundHandler | undefined
}

interface RouterRoute {
  path: string
  method: string
 // `handler: Handler` will be below:
  handlerContainer: HandlerContainer
}
```

**Comment by usualoma** [comment]
Update : https://github.com/honojs/hono/issues/851#issuecomment-1408506048

> <del>I don't think "not found" should be related to the handler.</del>
> Although the handler and notFound may be related, it is necessary to look for notFound outside of the handler.

**Comment by usualoma** [comment]
"not found" needs a little more thought, but I think the following is an approach that could be considered.
#854

**Comment by yusukebe** [comment]
@usualoma 

It seems to be good! This will not cause performance degradation.

About "not found". We can create the middleware handling `404` and return the custom response.

```ts
route.all('*', async (c, next) => {
  await next()
  if (c.res.status === 404) {
    return c.text('Not Found from sub', 404)
  }
})
```

This will work, right?

It's exactly what the Middleware-based structure helps to give us. I think your PR #854  and this handling will answer the requirement, what do you think? In that PR, though, we have to make sure to put the Error object in `c.error`.

**Comment by usualoma** [comment]
@yusukebe 
Roger, I understand about the 404 and `c.error`.
I will not have time this week, so I hope to improve #854 after the weekend.

**Comment by yusukebe** [comment]
@usualoma 

Thanks!

**Comment by nanderLP** [comment]
Thank you both for the quick response 😊

**Comment by usualoma** [comment]
@yusukebe 
I thought about the 404 for a while and reworked the PR.
How about a solution like #866?

**Comment by yusukebe** [comment]
@usualoma 

Thank you. I will reply later with other PR.

**Comment by usualoma** [comment]
@yusukebe 
I have tried to create two PRs to solve the problem of https://github.com/honojs/hono/pull/871#issuecomment-1422647262.

* #874
* #875

**Comment by yusukebe** [comment]
Hi @nanderLP !

Thanks to @usualoma , this issue has been fixed. I've released "v2.7.7". Try it. Thank you for creating the issue.
